chordsymbol (v5)

rvilarl opened the pull request:
This is the first complete example of vexflow 5. Each `ChordSymbolBlock` is now an `Element` and we use the `Element` infrastructure to draw them: all at the same location with their associated `xShift` or `yShift`. I believe that this has reduced the complexity of the code a lot (almost half of lines) and we can now use other fonts where we do not have the associated metrics.

rvilarl commented:
@ronyeh @AaronDavidNewman looking forward to get your feedback/reviews. 

There are visual differences.

rvilarl commented:
I have set this PR to draft to wait for the merge of #18. I will rebase then.

ronyeh commented:
Can you summarize what the visual differences are and why they are introduced with this PR?

Giving us the context / summary of the big idea would help. And if you can direct us to the correct files / methods and line numbers to start our review, that would help immensely!

AaronDavidNewman commented:
The symbols only appear in Petaluma font. Am I doing something wrong?

![image](https://github.com/vexflow/vexflow/assets/5438280/066fe121-f53d-42fe-bbf8-4304ac7cfb69)

![image](https://github.com/vexflow/vexflow/assets/5438280/b6dea2cf-059b-4dfb-9a99-f4656de2f2da)

rvilarl commented:
> The symbols only appear in Petaluma font. Am I doing something wrong?

@AaronDavidNewman this is the boxy problem that @mscuthbert refered to. I forgot to update the the load of web fonts, I have pushed it. Flow.html should work now fine but to get the testing infrastructure running you might need to install the fonts in your system. Please check how it goes.

rvilarl commented:
Rebased and ready to review

rvilarl commented:
> Can you summarize what the visual differences are and why they are introduced with this PR?

The source of the visual differences are two:
- The browser is positioning now main of the characters and glyphs, they are no longer calculated by us at run time.
- ChordSymbol is now using the appropriate Text Font for each family ie.: BravuraText for Bravura,... see HACK in flow.ts
 
> Giving us the context / summary of the big idea would help. And if you can direct us to the correct files / methods and line numbers to start our review, that would help immensely!

The big idea:
Each `ChordSymbolBlock` is now an `Element` and we use the `Element` infrastructure to draw them: all at the same location with their associated `xShift` or `yShift`. I believe that this has reduced the complexity of the code a lot (almost half of lines) and we can now use other fonts where we do not have the associated metrics.

This means:
- `SymbolTypes` is no longer required, all `ChordSymbolBlock` are `text`
- `IsSuperscript`, `IsSuperscript` are now members of `ChordSymbolBlock`
-  Font specific `ChordSymbolMetrics` and `TextFormatter` are no longer required. `ChordSymbolBlock` as an extension of `Element` knows how to calculate `metrics` and `width`. `yShift` and `xShift` is also provided by `Element`
- Kerning calculations no longer required.
- `Draw` is much simpler now calling `symbol.renderText`

Hope this helps!
-

ronyeh commented:
Thanks!

For a future reader dropping by:
- ChordSymbols are the elements that look like `C#7` that float above the stave, indicating which chord to play.
- It looks like this: https://unpkg.com/vexflow@4.2.2/tests/flow.html#filter=ChordSymbol


@rvilarl What do you use **Bravura Text** for? Is it just for the music symbols (like sharp ♯ and flat ♭) or are you using it to style the alpha numeric characters as well?

**Bravura Text** is not for styling alpha numeric characters. Bravura Text and regular Bravura are both used to render music symbols only.

See: https://github.com/steinbergmedia/bravura/blob/master/redist/bravura-text.md

Maybe the glyph registration points are different than regular Bravura, and maybe each glyph's **advance width** is customized so that the font is easier to use in a word processor.

Bravura did not come with a special font for displaying regular text alongside Bravura scores. Instead, the designers recommend some text fonts in their [metadata.json](https://github.com/steinbergmedia/bravura/blob/02e8ed29a29115df35007d1178cebaeee26c20e1/redist/bravura_metadata.json#L28-L33) file:

https://github.com/steinbergmedia/bravura/blob/02e8ed29a29115df35007d1178cebaeee26c20e1/redist/bravura_metadata.json#L28-L33

```json
"textFontFamily":[
  "Academico",
  "Century Schoolbook",
  "Edwin",
  "serif"
],
```


Compare this with Petaluma: https://github.com/steinbergmedia/petaluma/tree/master/redist/otf

Petaluma also has a PetalumaText. Both are for displaying music symbols.

However, since Petaluma is supposed to look like handwriting, the designers also included **PetalumaScript** for displaying alphanumeric characters alongside scores engraved with Petaluma.


When @AaronDavidNewman contributed the ChordSymbol work in 2020, he and Mohit had a discussion about possible fonts to pair with the score:

https://github.com/0xfe/vexflow/pull/790#issuecomment-634589783
https://github.com/0xfe/vexflow/pull/790#issuecomment-637512786

I think they eventually arrived on Roboto Slab.

In the end, it's personal preference. But now that @sschmidTU has alerted us to the possible GDPR compliance problems with using Google Web Fonts (or any non-self-hosted font), perhaps it is easier if we just paired Bravura with with a web safe font that is included with most operating systems? 

We could use the font stack:
```
Georgia, Times, "Times New Roman", serif;
```

https://www.cssfontstack.com/Georgia
https://www.cssfontstack.com/Times-New-Roman

Then we can teach the developer how they can customize the font if they wish, using our vexflow-fonts or using google fonts, or by self hosting fonts.

What do you all think? Should we remove Roboto Slab from VexFlow by default, as long as we make it easy for the developer to self host or integrate their favorite web fonts?

ronyeh commented:
Here is a discussion about mixing regular text alongside Bravura symbols, with comments from the original designer:

https://forums.steinberg.net/t/ot-how-to-use-bravura-text/160206/6

ronyeh commented:
Here the Bravura designer mentions that Academico is also available under SIL open font license:

https://forums.steinberg.net/t/academico-open-font-license/114847

So maybe we should just extract the Academico font from Dorico and include it in vexflow-fonts.

ronyeh commented:
Academico: https://befonts.com/academico-serif-font.html

rvilarl commented:
Hi @ronyeh Thanks a lot for looking into that! The current HACK in flow takes the following combinations:
```js
      case 'Bravura':
        CommonMetrics.fontFamily = 'Bravura,BravuraText';
        break;
      case 'Gonville':
        CommonMetrics.fontFamily = 'GonvilleSmufl,Bravura,Roboto Slab';
        break;
      case 'Leland':
        CommonMetrics.fontFamily = 'Leland,Bravura,LelandText';
        break;
      case 'Petaluma':
        CommonMetrics.fontFamily = 'Petaluma,Bravura,PetalumaScript';
```
It would be great to include Academico in vexflow-fonts! What should we use with Leland. I will also make a PR now with an improved inspector of the glyphs.

ronyeh commented:
<img width="988" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/a7905bc8-9ec3-4f20-af30-4d59e08dd001">

Here I demonstrate that Bravura Text doesn't include any glyphs for ASCII characters. The font family I used was:

```
font-family: 'Bravura Text', Impact;
```

And the HTML characters are:

```
<div> - &#xE050; - C&#x266D; - G&#x266E; - Em&#x266F;7</div>
```

The first character is just copy and pasted directly from the SMuFL specification webpage, so it'll probably show a box. The other music symbols use the unicode code points.

ronyeh commented:
<img width="978" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/d1b16fd2-6053-4423-be8c-5321b4e68337">

```
font-family: 'Bravura Text', 'Academico';
```

rvilarl commented:
But I would rather use the combination

```css
font-family: 'Bravura, Academico';
```

`Bravura Text` is bigger, and looking into the details, it contains the same glyphs repeated at different yOffsets.

rvilarl commented:
I have read that the only problem with the google fonts is to download them from google because google stores the IP address of the client using the font. If we download it from vexflow-fonts it should not be a problem, I believe.

ronyeh commented:
<img width="1165" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/e670e159-e5d2-4ab1-a323-e18ae5b42061">

Here is the difference when Bravura Text or Bravura are used alongside ASCII characters typeset in Academico.

Bravura Text is aligned properly when placed next to regular text. 

Plain old Bravura glyphs might take up more vertical space, and have different glyph origins.

rvilarl commented:
@ronyeh you are not using the right glyphs for the accidentals in the chordsymbols, you need these

https://w3c.github.io/smufl/latest/tables/standard-accidentals-for-chord-symbols.html?highlight=csym#standard-accidentals-for-chord-symbols-ued60ued6f

They are already used in this PR

ronyeh commented:
> I have read that the only problem with the google fonts is to download them from google because google stores the IP address of the client using the font. If we download it from vexflow-fonts it should not be a problem, I believe.

I do not know what unpkg.com does. That is not a server we control. Like jsdelivr, unpkg just makes NPM packages available over the web. We uploaded some SMuFL fonts to it so that we can load web fonts with VexFlow.

To be 100% GDPR safe, the developer would most likely need to self-host their fonts.

I have started experimenting with this. At the moment, you can npm install individual woff2 fonts as follows:

```
npm install @vexflow-fonts/bravura
npm install @vexflow-fonts/bravura-text
```

If this approach works for people, we can also upload Academico or other SIL open fonts to NPM, so that they can be installed individually. I borrowed this approach from [fontsource.org](https://fontsource.org/).

ronyeh commented:
> @ronyeh you are not using the right glyphs for the accidentals in the chordsymbols, you need these
> 
> https://w3c.github.io/smufl/latest/tables/standard-accidentals-for-chord-symbols.html?highlight=csym#standard-accidentals-for-chord-symbols-ued60ued6f
> 
> The hieght of the box is resolved using actualBoundingBoxAscent/Descent metrics rather than fontBoundingBoxAscent/Descent.
> 
> They are already used in this PR

OK, I was just using the glyph code points recommended by the original font designer:

https://forums.steinberg.net/t/ot-how-to-use-bravura-text/160206/6

<img width="775" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/01a0715d-fa50-4083-9914-9ddd2a21f8cc">


In this example, I was typesetting the sharp and flat symbols next to regular Academico text. It is possible that we want to use different glyph code points when placing sharp and flat symbols next to other music symbols (note heads).

ronyeh commented:
I read your link. Thanks! I do see that there are special sharp and flat glyphs for use alongside chord symbols. 

I think the tricky part is what happens when chord symbols and sharp/flat symbols are used next to Academico text?

AaronDavidNewman commented:
5 more responses came in while I was typing.  So I'll keep it short.

glyphs need to be scaled up, they are too small.  So you will need to add the distinction back between 'text' and 'glyph', even if they come from the same font file since they will need to be in separate elements.

unpkg method won't work with local development.  You need to include the static fonts you want it to use, or put the woff in a cdn and reference them statically in the html file.  Otherwise you will be fielding questions forever about why xxx doesn't show up in my computer.

Other than the issue with loading fonts and the scale, I think this scheme will work for chord symbols.

ronyeh commented:
> unpkg method won't work with local development. You need to include the static fonts you want it to use, or put the woff in a cdn and reference them statically in the html file. Otherwise you will be fielding questions forever about why xxx doesn't show up in my computer.

Thanks @AaronDavidNewman 

Can you clarify a abit? When you say put the woff in a CDN.... that's what we are trying to do with the unpkg.com (which is a CDN for serving NPM packages).

For example, here is the Bravura woff2: https://unpkg.com/vexflow-fonts@1.0.5/bravura/Bravura_1.392.woff2

You can use the **@font-face** CSS declaration and point to that URL.

VexFlow also provides a JS way to do the same thing. The FontFace JS API is equivalent to the CSS `@font-face` method.

`Font.loadWebFont(fontName: string, woffURL: string, includeWoff2?: boolean)`

```
  static async loadWebFont(fontName: string, woffURL: string, includeWoff2: boolean = true): Promise<FontFace> {
    const woff2URL = includeWoff2 ? `url(${woffURL}2) format('woff2'), ` : '';
    const woff1URL = `url(${woffURL}) format('woff')`;
    const woffURLs = woff2URL + woff1URL;
    const fontFace = new FontFace(fontName, woffURLs);
    await fontFace.load();
    document.fonts.add(fontFace);
    return fontFace;
  }
```

Regarding glyph vs text sizes, I think that's why the sizing and registration points are different in Bravura Text. When symbols are placed directly next to ASCII text, they need to be aligned and sized correctly.

But if we want to just use Bravura (my preference, so we don't have to load multiple fonts), then we need to figure out the alignment and sizing so it looks correct.

ronyeh commented:
See here for description of the FontFace API: https://developer.mozilla.org/en-US/docs/Web/API/CSS_Font_Loading_API

I believe it is equivalent to the `@font-face` blocks in CSS, except since it is in JS, we can control exactly when the fonts are loaded.

AaronDavidNewman commented:
> Can you clarify a bit? When you say put the woff in a CDN

I should have specified.  Maybe it will work but I think pushing an https script into javascript won't work in http context for security reasons.  But I could be wrong.

Anyway, I am getting this error:


```
Request URL:
https://unpkg.com/vexflow-fonts@1.0.5/bravura/Bravura_1.392.woff22
Request Method:
GET
Status Code:
404
Remote Address:
104.16.126.175:443
Referrer Policy:
strict-origin-when-cross-origin
```

AaronDavidNewman commented:
I fixed the error with the additional '2' (`  static async loadWebFont`) and I can now load the fonts.   But the Bravura text font is not getting used, it uses Times New Roman instead (which I can tell from chrome DOM inspector). 

Does this work for others?  I don't see how it would if Bravura doesn't have the ascii symbols, like Ron says.   No 'F7' b/c no 'F'.

rvilarl commented:
> I fixed the error with the additional '2' (` static async loadWebFont`) and I can now load the fonts. But the Bravura text font is not getting used, it uses Times New Roman instead (which I can tell from chrome DOM inspector).

Sorry for the 2, now updated.

> Does this work for others? I don't see how it would if Bravura doesn't have the ascii symbols, like Ron says. No 'F7' b/c no 'F'.

As @ronyeh  said Bravura / BravuraText combination makes no sense. I think that this is working fine for the Petaluma / PetalumanScript combination.

AaronDavidNewman commented:
> As @ronyeh said Bravura / BravuraText combination makes no sense

OK, then maybe we can move it back to Robot Slab for non-handwritten fonts.  I thought that worked the best with the chord changes, and it is a pretty complete font wrt different character sets, weights, styles etc.

AaronDavidNewman commented:
One more comment for the actual code part.  In your code, you space the blocks like so:

``` typescript
width += block.getWidth() + ChordSymbol.minPadding;
```
The old code had this:
``` typescript
block.width += ChordSymbol.spacingBetweenBlocks * fontSize * superSubScale;
```

This is why the chord changes are all squished together in the v5 version.  Since you put text and glyph in the same block, there is no spacing between them and so the symbols are both too small and too close to the text.  And the spacing needs to be scaled to the fonts size/super/sub so the spacing is the same in the superscripts.

V5:
![image](https://github.com/vexflow/vexflow/assets/5438280/12d525a1-2606-40a2-b5f5-526530131d8d)

vs.
![image](https://github.com/vexflow/vexflow/assets/5438280/07505fdf-c4c0-4d69-9d8c-9f0cf5208e2a)

AaronDavidNewman commented:
Another code comment: you left in the kerning tests, but removed the kerning logic.  

v5:
![image](https://github.com/vexflow/vexflow/assets/5438280/dca4bb6b-ce30-4ff5-8efe-e638a42ea31e)

v4:
![image](https://github.com/vexflow/vexflow/assets/5438280/8c618a57-458e-45bf-98c8-a8954dd267ef)

ronyeh commented:
@rvilarl For the purposes of this PR, can we go back to Roboto Slab just so we can get it as close as possible (visually) to VexFlow 4.x? It'll make things easier to reason about if we aren't changing too many things at the same time.

The big idea of this PR is how we draw & position glyphs, not that which text font we should pair with Bravura. In the future, we can revisit this and pick one of the following:
1. Have default VexFlow be completely web safe with regard to the text fonts.... so we would use boring fonts like Arial, Times New Roman, Georgia, for the ChordSymbols that are paired with the Bravura glyphs.
2. Use a web font and load it from CDN. Academico is the preferred text font as suggested by Steinberg designers. I'm not sure there is a woff2 available, as my copy of Dorico only has the OTFs. I think it is pretty easy to convert OTF => WOFF2 though.
3. Use a web font but install it via `npm install <font-name>`. For example, **fontsource** has Roboto Slab up on NPM: https://fontsource.org/fonts/roboto-slab/install  If we don't want to rely on **fontsource**'s npm packages, we could always roll our own npm packges (e.g., `npm install @vexflow-fonts/roboto-slab` or `npm install @vexflow-fonts/academico`).

Installing the font via npm install is an interesting idea, because you only need to be online to clone the vexflow repo and to do the NPM install. Afterward, you could work on a plane trip and your fonts should still show up correctly (as long as your dev server and bundler knows how to load the font).

Anyways, back to this PR. I vote for trying to minimize the visual differences as much as possible (so maybe we go back to Roboto Slab for now).

We should discuss kerning and other issues. I don't mind temporarily breaking some things, as long as we believe we have a possibility / path to restore those features before we ship 5.0 for real.

ronyeh commented:
> > I fixed the error with the additional '2' (` static async loadWebFont`) and I can now load the fonts. But the Bravura text font is not getting used, it uses Times New Roman instead (which I can tell from chrome DOM inspector).
> 
> Sorry for the 2, now updated.


Hi @rvilarl:

I wrote the original `loadWebFont()` function, and I do not believe it needs to be updated (to remove the "extra 2").

Here is the original version:

https://github.com/0xfe/vexflow/blob/master/src/font.ts#L350-L362

```ts
static async loadWebFont(fontName: string, woffURL: string, includeWoff2: boolean = true): Promise<FontFace> {
    const woff2URL = includeWoff2 ? `url(${woffURL}2) format('woff2'), ` : '';
    const woff1URL = `url(${woffURL}) format('woff')`;
    const woffURLs = woff2URL + woff1URL;
    // eslint-disable-next-line
    // @ts-ignore
    const fontFace = new FontFace(fontName, woffURLs);
    await fontFace.load();
    // eslint-disable-next-line
    // @ts-ignore
    document.fonts.add(fontFace);
    return fontFace;
}
```

The idea is that you pass in the woffURL, which points to a file named `my-awesome-font.woff`. By default, we will append the number `2` to the end of the *.woff URL, so we end up with `my-awesome-font.woff2`. Then we will provide BOTH woff2 and woff URLS to the `FontFace()` constructor. That's how web fonts are traditionally specified, with both a woff2 and a woff URL.

However, 96% of browsers support woff2 now, so if you really want to drop the support for the woff2 + woff approach that I wrote, I'm OK with that. We just need to make sure existing tests and demos don't break.

But I think you should be able to use my original `loadWebFont(...)` without modification. Just make sure you don't add the 2 at the end of the URLs you provide to the function.

ronyeh commented on src/font.ts:
```diff
@@ -326,16 +324,34 @@ export class Font {
   static WEB_FONT_FILES: Record<string /* fontName */, string /* fontPath */> = {
-    'Roboto Slab': 'robotoslab/RobotoSlab-Medium_2.001.woff',
-    PetalumaScript: 'petaluma/PetalumaScript_1.10_FS.woff',
+    Bravura: 'bravura/Bravura_1.392.woff2',
+    BravuraText: 'bravura/BravuraText_1.393.woff2',
+    GonvilleSmufl: 'gonvillesmufl/GonvilleSmufl_1.100.woff2',
+    Gootville: 'gootville/Gootville_1.3.woff2',
+    GootvilleText: 'gootville/GootvilleText_1.2.woff2',
+    Leland: 'leland/Leland_0.75.woff2',
+    LelandText: 'leland/LelandText_0.75.woff2',
```
Note how the two deleted lines ended in *.woff.

There is no need to specify the woff2 file extension, because the loadWebFonts() function will automatically add the woff2 extension before providing it to the FontFace API.

rvilarl replied:
I decided to go explicitely to woff2 because there was a lot of discussion on network traffic and total length of files.

ronyeh commented on src/flow.ts:
```diff
@@ -228,8 +221,24 @@ export class Flow {
+    // HACK during review to introduce the correct stacks step by step
+    switch (fontNames[0]) {
+      case 'Bravura':
+        CommonMetrics.fontFamily = 'Bravura,BravuraText';
```
Yeah, we probably shouldn't use both Bravura and BravuraText, as they contain the same glyph code points.

rvilarl replied:
Roboto Slab already there. :)

ronyeh commented on src/flow.ts:
```diff
@@ -228,8 +221,24 @@ export class Flow {
+        CommonMetrics.fontFamily = 'GonvilleSmufl,Bravura,Roboto Slab';
+        break;
+      case 'Leland':
+        CommonMetrics.fontFamily = 'Leland,Bravura,LelandText';
```
Same here. Drop LelandText from the font stack.

rvilarl replied:
done

ronyeh commented on src/flow.ts:
```diff
@@ -228,8 +221,24 @@ export class Flow {
+        CommonMetrics.fontFamily = 'Bravura,BravuraText';
+        break;
+      case 'Gonville':
+        CommonMetrics.fontFamily = 'GonvilleSmufl,Bravura,Roboto Slab';
```
This is interesting and unexpected for me. The font stack includes two music fonts and one text font.

I guess that sort of makes sense... if our ChordSymbol is a text string that includes a mix of music symbols and ASCII text symbols?

rvilarl replied:
It is indeed expected. The approach in Vexflow5 is to specify all the fonts and the browser will fall back to the next one when the glyph is not defined. We search in SMuFL first and then in standard text. My fault was to believe that BravuraText was a standard text file to combine with Bravura.

ronyeh reviewed (changes requested):
This is just a preliminary review. I haven't had the chance to read through every change yet.

I don't think loadWebFonts() needs to be modified at all. We just need to fix the URLs in WEB_FONT_FILES to not include the 2 at the end. Each file should end in `.woff`

ronyeh commented:
It looks like in Dorico, they use Bravura Text for chord symbols:

https://steinberg.help/dorico/v2/en/dorico/topics/notation_reference/notation_reference_chord_symbols_appearance_font_changing_t.html

> We recommend that you do not edit Chord Symbols Music Text Font, which is set to Bravura Text by default. This can only be set to a SMuFL-compliant font intended for use in text-based applications.

So I guess the app uses Bravura for the symbols on the staff, and Bravura Text for any symbols that happen in chord symbols, lyrics, and other places that place music symbols next to regular alpha numeric text.

ronyeh commented:
https://w3c.github.io/smufl/latest/tables/chord-symbols.html?highlight=chord#chord-symbols-ue870ue87f

At the bottom of the linked page:

> Scoring applications should be able to create strings with complex formatting, e.g. superscript and subscript characters, small digits stacked on top of each other, and scale these symbols to any arbitrary size in order to produce satisfactory chord symbols with a wide variety of visual appearances.

So it's the responsibility of the scoring application (Dorico... or in our case, VexFlow) to decide how to scale and kern the glyphs to make satisfactory looking chord symbols.

rvilarl commented:
@ronyeh @AaronDavidNewman please be patient, I am looking into option to improve the quality of the ChordSymbols taking into account the comments from both of you. Some notes:
- it is possible to make a single SVG Text that represent the complete symbol with sub/super indexes and different font sizes maiking use of tspan. However I do not see something similar for the canvas.
- the space between characters can be adjusted both in canvas and svg text with `letter-spacing`

@ronyeh could you create a vexflow-fonts with the text fonts suggested for Bravura, Leland, Gonville & Petaluma? One was Academico (for Bravura) and thanks for all your investigations on fonts!

rvilarl commented:
I have restored robot slab in all but petaluma, where I understood that the comibantion with PetalumaScript was fine.

rvilarl commented on tests/chordsymbol_tests.ts:
```diff
@@ -69,7 +61,7 @@ function withModifiers(options: TestOptions): void {
     formatter.joinVoices([voice]);
     const voiceW = formatter.preCalculateMinTotalWidth([voice]);
-    const staffW = voiceW + Stave.defaultPadding + getGlyphWidth('gClef');
+    const staffW = voiceW + Stave.defaultPadding + getGlyphWidth('E050' /*gClef*/);
```
This is the approach. I have done it in src but I missed to look into tests.

rvilarl commented on src/font.ts:
```diff
@@ -326,16 +324,34 @@ export class Font {
    * Or you can use your own host.
    */
-  static WEB_FONT_HOST = 'https://unpkg.com/vexflow-fonts@1.0.3/';
+  static WEB_FONT_HOST = 'https://unpkg.com/vexflow-fonts@1.0.5/';
```
I agree. It is specially inconvenient whe you have no internet connection. I had to disable that several times flying.

ronyeh replied:
I have been making npm packages for each individual font. Once installed via npm install, they should be available for local use in our project. Eventually, we will be able to support offline mode (e.g., coding on an airplane).

https://www.npmjs.com/settings/vexflow-fonts/packages

rvilarl commented on tests/font_tests.ts:
```diff
@@ -53,17 +53,17 @@ function setFont(assert: Assert): void {
   // Add italic to the default font as defined in Element.TEXT_FONT (since Accidental does not override TEXT_FONT).
   flat.setFont(undefined, undefined, undefined, 'italic');
-  assert.equal(flat.getFont(), 'italic 30pt Bravura,Gonville,Custom');
+  assert.equal(flat.getFont(), 'italic 30pt Bravura,BravuraText');
```
Roboto Slab is now there.

rvilarl commented on src/font.ts:
```diff
@@ -326,16 +324,34 @@ export class Font {
+    PetalumaScript: 'petaluma/PetalumaScript_1.10_FS.woff2',
+    MuseJazz: 'musejazz/MuseJazz_1.0.woff2',
+    MuseJazzText: 'musejazz/MuseJazzText_1.0.woff2',
+    'Roboto Slab': 'robotoslab/RobotoSlab-Medium_2.001.woff2',
```
changed

rvilarl commented:
I had to revert the renaming. If you install this google font on your system, it is register as 'Roboto Slab' and the images generated on the tests do not get the appropriated font. This was happening with 'Petaluma Script' as well.

rvilarl commented:
I did not find a way similar to tspan in canvas. The current bravura output is like this:

![pptr-ChordSymbol Top_Chord_Symbols Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/5363a582-faf8-4768-8a56-e12de89cbcd8)
![pptr-ChordSymbol Bottom_Chord_Symbols Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/f257b09b-7488-46b6-8d68-d710c8497e14)
![pptr-GlyphNote GlyphNote_with_ChordSymbols Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/c2b93ca1-c9bb-4c82-a2a9-e8344aa085d9)
![pptr-ChordSymbol Top_Chord_Symbols_Justified Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/9f78d404-ee46-4ab6-8a93-315ae610109b)
![pptr-ChordSymbol Double_Bottom_Chord_Symbols Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/94977214-550e-4bcc-89ee-13e4faf39b60)
![pptr-ChordSymbol Chord_Symbol_With_Modifiers Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/9e923cff-0f00-4433-94c7-3d1508a6a27e)
![pptr-ChordSymbol Chord_Symbol_Kerning_Tests Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/0a5b5015-8939-415c-9ea4-945087d15949)
![pptr-ChordSymbol Chord_Symbol_Font_Size_Tests Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/8d38a89f-24a9-4d67-8a73-1b3581e044a3)
![pptr-ChordSymbol Bottom_Stem_Down_Chord_Symbols Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/64df08a6-6fa5-494d-893d-a3a91d8a1ef9)

rvilarl commented:
Petaluma outputs (here the kerning is not so important as A is a bit rotated):

![pptr-ChordSymbol Top_Chord_Symbols Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/d0942fed-a902-4138-b67b-4c26b48dd8f9)
![pptr-ChordSymbol Bottom_Chord_Symbols Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/2f7e0652-e840-4d14-8a6f-5612fb821a6d)
![pptr-GlyphNote GlyphNote_with_ChordSymbols Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/b9f779b0-e136-4d7c-a4c8-df84a5b7685f)
![pptr-ChordSymbol Top_Chord_Symbols_Justified Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/8e90222d-8bcf-407f-a9d0-d45ae791b73e)
![pptr-ChordSymbol Double_Bottom_Chord_Symbols Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/e639e471-7c7e-4b93-8fe4-f63b15ab7cc9)
![pptr-ChordSymbol Chord_Symbol_With_Modifiers Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/94892d27-498e-49c2-a71e-028af5f5c66d)
![pptr-ChordSymbol Chord_Symbol_Kerning_Tests Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/542cadfe-4960-4bbb-8387-3df5786033e4)
![pptr-ChordSymbol Chord_Symbol_Font_Size_Tests Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/a770825f-d856-40c2-b448-b5ae1c18a3de)
![pptr-ChordSymbol Bottom_Stem_Down_Chord_Symbols Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/e6c8aa58-6262-429d-a162-726d6aaa7c9b)

rvilarl commented:
@AaronDavidNewman @mscuthbert @ronyeh what has to be fixed to make them acceptable?

ronyeh commented:
I think my main concern is if we do not have a (plausible) future solution for kerning, then we will be removing a feature that some folks (at least Aaron) like. :-)

If you think we have a possibility to reintroduce kerning for these symbols at some point (and can outline a rough idea as to how we might try to do it), then I would be okay with moving forward, even if it breaks kerning for now.

ronyeh commented:
In your most recent screenshots, I actually think the parentheses glyphs are too close to the other symbols. It feels too crowded to me. Are we over-kerning? In a way, if we removed kerning, everything would be slightly more spaced out and less crowded. Maybe that's good for some cases?

For example, in this screenshot you upoaded: https://user-images.githubusercontent.com/22865285/253409138-5363a582-faf8-4768-8a56-e12de89cbcd8.png

I think the # and b symbols are too close to the 11 and 9, respectively. I also think the parentheses are too close to the symbols as well, except for the one on the second row left side. That looks like kerning is turned off (and I actually kind of prefer the more spaced out look).

ronyeh commented:
> I had to revert the renaming. If you install this google font on your system, it is register as 'Roboto Slab' and the images generated on the tests do not get the appropriated font. This was happening with 'Petaluma Script' as well.

Ah yes... I agree it is a good idea to follow the original naming as defined by the font designer.

Just for the record.... do both 'Roboto Slab' has a space, and 'Petaluma Script' need a space? For some reason I thought PetalumaScript did NOT have a space inside.

We can verify by dragging font files onto this website to inspect the name: 

https://fontdrop.info/

Or you can also load it into the opentype.js website.

https://opentype.js.org/font-inspector.html


So... I think the confusing part is that there are multiple fields for a font's name.

If I inspect the Roboto Slab OTF, I get a windows naming table that says:

<img width="1139" alt="image" src="https://github.com/vexflow/vexflow/assets/239113/454e06ab-0744-457c-8bf9-b1be608c48e0">

The postScriptName is `RobotoSlab-Medium` as PS does not allow spaces. The preferredFamily is `Roboto Slab`, with a space inside.


If I inspect the Petaluma Script OTF, I get:
```
fontFamily = Petaluma Script
fontSubfamily = Regular
fullName= PetalumaScript
postScriptName = PetalumaScript
```

So even though the postscriptName never has a space, I think we should be consistent and go with whatever is specified under `fontFamily`.

ronyeh commented on src/font.ts:
```diff
@@ -347,13 +363,8 @@ export class Font {
-    // @ts-ignore
-    const fontFace = new FontFace(fontName, woffURLs);
+  static async loadWebFont(fontName: string, woffURL: string): Promise<FontFace> {
+    const fontFace = new FontFace(fontName, `url(${woffURL})`);
```
If you look at the original version of loadWebFont, it specifies the woff2URL first, and then the woff1URL. The constructor accepts fallback fonts: https://stackoverflow.com/questions/41061599/how-to-load-fallback-font-formats-with-fontface-api

The browser will look for the woff2 file first, and if it returns a 404, it will fall back to the woff file. So this shouldn't have an impact on network traffic if the woff2 file exists and the user's browser supports woff2.

Anyways, I am OK if you change the API here since the original isn't the clearest. 

However, it would be nice if there was a way to support the fallback approach. The FontFace constructor does not require the url to point to a woff file. In fact, I've verified that you can pass in an OTF and my version of Chrome loads the OTF perfectly fine. :-)

Maybe the method signature here should be:

```
static async loadWebFont(fontName: string, fontURLs: string | string[]): Promise<FontFace> {
```

If we pass in a single string, then do what you are doing now. 

If we pass in a string[], then build up a comma separated list of fallback URLs?

rvilarl replied:
What about providing the name without extension?

```
  static WEB_FONT_FILES: Record<string /* fontName */, string /* fontPath */> = {
    Bravura: 'bravura/Bravura_1.392',
    BravuraText: 'bravura/BravuraText_1.393',
...
 ```

and then adding them in the function.

ronyeh replied:
OK... I threw together (an incomplete and untested) approach for `loadWebFont()`

```ts
static async loadWebFont(fontFamily: string, fontURLs: string | string[], localNames?: string | string[]) {
    // Compose a src string that looks like:
    // local('FontName'), local('FontName-Bold'), url('font.woff2') format('woff2'), url('font.woff') format('woff')
    let sources: string[] = [];

    if (localNames) {
        if (typeof localNames === "string") {
            sources.push(`local('${localNames}')`);
        } else {
            for (const n of localNames) {
                sources.push(`local('${n}')`);
            }
        }
    }

    if (typeof fontURLs === "string") {
        fontURLs = [fontURLs];
    }

    for (const fontURL of fontURLs) {
        // Determine the type based on the extension.
        // .otf => opentype
        // .ttf => truetype
        // .woff => woff
        // .woff2 => woff2
        const extension = fontURL.split(".").pop();
        let fontType = "";
        switch (extension) {
            case "otf":
                fontType = "opentype";
                break;
            case "ttf":
                fontType = "truetype";
                break;
            case "woff":
                fontType = "woff";
                break;
            case "woff2":
            default:
                fontType = "woff2";
                break;
        }

        sources.push(`url('${fontURL}') format('${fontType}')`);
    }

    const src = sources.join(", ");
    console.log(src);
    // ... DO THE REST HERE
    // new FontFace(fontFamily, src).....
}

```

ronyeh replied:
This new approach allows developers to load ONE or MORE fonts. The font types will be automatically detected by file extension.

Additionally, the developer can specify a local name (or more than one) so that the user agent can first look on the local system and load that font (without making a network request).

rvilarl replied:
I suggest tha you make a PR for this one

rvilarl commented:
What about using the text font parenthesis?

![pptr-ChordSymbol Chord_Symbol_With_Modifiers Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/8a203337-4856-47d7-8b69-964f0bb798bf)
![pptr-ChordSymbol Chord_Symbol_With_Modifiers Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/927ce453-7548-4fe3-b808-f900109be3ca)

ronyeh commented:
Good idea. The only edge case that wouldn't solve is if the developer
prefers to load the OTF instead. We can always go with your approach for
now. Just rename it fontURL instead of woffURL. Later on we can accept a
string[] as well. We can do that in a separate PR.

On Thu, Jul 13, 2023, 9:51 PM Rodrigo Vilar ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/font.ts
> <https://github.com/vexflow/vexflow/pull/15#discussion_r1263286855>:
>
> > @@ -347,13 +363,8 @@ export class Font {
>    // Support distributions of the typescript compiler that do not yet include the FontFace API declarations.
>    // eslint-disable-next-line
>    // @ts-ignore
> -  static async loadWebFont(fontName: string, woffURL: string, includeWoff2: boolean = true): Promise<FontFace> {
> -    const woff2URL = includeWoff2 ? `url(${woffURL}2) format('woff2'), ` : '';
> -    const woff1URL = `url(${woffURL}) format('woff')`;
> -    const woffURLs = woff2URL + woff1URL;
> -    // eslint-disable-next-line
> -    // @ts-ignore
> -    const fontFace = new FontFace(fontName, woffURLs);
> +  static async loadWebFont(fontName: string, woffURL: string): Promise<FontFace> {
> +    const fontFace = new FontFace(fontName, `url(${woffURL})`);
>
> What about providing the name without extension?
>
>   static WEB_FONT_FILES: Record<string /* fontName */, string /* fontPath */> = {
>     Bravura: 'bravura/Bravura_1.392',
>     BravuraText: 'bravura/BravuraText_1.393',
> ...
>
> and then adding them in the function.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vexflow/vexflow/pull/15#discussion_r1263286855>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB2MCOIOALOM4VN7ZN5NY3XQDF33ANCNFSM6AAAAAA2EXA7S4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

AaronDavidNewman commented:
> What about using the text font parenthesis?

How does it look?  'veryTall' parenthesis are supposed to be taller than 'tall' parenthesis.  They also have different placement (1/2 below the baseline).  I think those are what we used for stacked chords (x/y).

ronyeh commented on demos/node/fonts/load.js:
```diff
@@ -2,7 +2,7 @@
 // This demo shows how to use node-canvas to load a font (woff / otf).
 // It does not currently use VexFlow.
-// TODO: Check if the text fonts (PetalumaScript & Roboto Slab) are automatically
+// TODO: Check if the text fonts (PetalumaScript & RobotoSlab) are automatically
```
Are we using spaces for both Petaluma Script and Roboto Slab?

rvilarl replied:
done

rvilarl commented:
> What about using the text font parenthesis?
> 
> ![pptr-ChordSymbol Chord_Symbol_With_Modifiers Bravura svg_current](https://github.com/vexflow/vexflow/assets/22865285/8a203337-4856-47d7-8b69-964f0bb798bf)
> ![pptr-ChordSymbol Chord_Symbol_With_Modifiers Petaluma svg_current](https://github.com/vexflow/vexflow/assets/22865285/927ce453-7548-4fe3-b808-f900109be3ca)
> 
> 
> 

@AaronDavidNewman this is how the text parentheses look.

AaronDavidNewman commented:
They are short/fat compared to the stacked scenario.  For the ones on the same line, it's not too noticeable.

I realize I am comparing a screen shot to an svg, but I'm just focused on the the formatting:

Images from above:
![image](https://github.com/vexflow/vexflow/assets/5438280/04a0ba6c-977e-4ef7-b7a8-9912131d3ac9)
![image](https://github.com/vexflow/vexflow/assets/5438280/feb481ed-08cf-415e-b956-1085c7503ca1)

VF 4:
![image](https://github.com/vexflow/vexflow/assets/5438280/78d2f730-1daa-4fc2-9b66-66cad4f1c9d8)
![image](https://github.com/vexflow/vexflow/assets/5438280/54c50655-e1f6-4b7d-bb64-7f30345811cd)

ronyeh commented:
I guess it's a matter of personal preference.

For the V5 images that Aaron cropped, I do agree the Petaluma parentheses seem rather thick. However, the Roboto Slab parentheses match well with the Roboto Slab font.

In the V4 images, I think the Bravura parentheses do not visually pair well with the Roboto Slab font. The Bravura parentheses are too thin in my opinion. Also, I think the 7 in the F7 is too close to the left parentheses in the last image (Roboto + Bravura).

I am fine with either solution.

In theory I would prefer the parentheses to be from the **text font**, because to me, it flows better visually (especially in the Roboto Slab example).

However, SMuFL defines a bunch of parentheses glyphs for chord symbols, so we might as well use them :-)
https://w3c.github.io/smufl/latest/tables/chord-symbols.html

I think in this case, Academico might fit better alongside Bravura, since the Roboto Slab variant that we are using has a heavier weight than the Bravura parentheses symbols.


Regarding this PR, I would lean towards keeping it as close as possible to the V4 look and feel.

In the interest of moving forward, I'd be okay if this PR did not include a solution for kerning, as long as we keep the issue open and can revisit it before shipping 5.0. We can always learn from the approach that Aaron took and see if we can bring some of those ideas over to 5.0.

ronyeh commented on src/flow.ts:
```diff
@@ -228,8 +221,24 @@ export class Flow {
    * @returns an array of Font objects corresponding to the provided `fontNames`.
    */
   static setMusicFont(...fontNames: string[]): Font[] {
+    // HACK during review to introduce the correct stacks step by step
```
So after this PR is merged, will this "HACK" go away? What is the correct solution in this method?

rvilarl replied:
The hack will stay until all the code is ported this hack avoids that the non ported files produce visual diffrences. In the end it will be:

```
static setMusicFont(...fontNames: string[]): void {
    // Convert the array of font names into an array of Font objects.
    CommonMetrics.fontFamily = fontNames.join(',');
  }
```

ronyeh commented on tools/fonts/inspector/otf.html:
```diff
@@ -214,7 +214,7 @@
         document.fonts.add(gonville);
         const robotoSlabOTF = '../@/robotoslab/RobotoSlab-Medium_2.001.otf';
-        const robotoSlab = new FontFace('Roboto Slab', `url(${robotoSlabOTF}) format('opentype')`);
+        const robotoSlab = new FontFace('RobotoSlab', `url(${robotoSlabOTF}) format('opentype')`);
```
Just checking if we are using spaces in all the font names. Is it supposed to be "Roboto Slab", with a space?

rvilarl replied:
done

AaronDavidNewman commented:
> In the interest of moving forward, I'd be okay if this PR did not include a solution for kerning

Yes, we should definitely move on.

I wonder now if it even makes sense to keep the chord symbols module.  The thing it offered in V4 is 1) the ability to mix text and glyphs, and 2) the ability to pass in your own metrics outside of VexFlow main build.  1 is no longer required, because we are using the glyphs/test directly from the font.  And 2 is no longer possible because there are no longer any metrics to use for formatting.

So the only thing in ChordSymbol that isn't also in annotation is super/subscript, which we could move to annotation at some future time.  It might be useful for other things.

If you want to keep it, you can probably just go with what you have.

ronyeh commented:
I'd vote to keep the chord symbols module for now. It makes it easier to
reason about our migration from v4 to v5, as we can compare features
module-to-module.

At some point once things are more stable, we might realize that
chordsymbol is largely redundant.... Then maybe it could be refactored as a
subclass or eliminated/merged into another class.

On Tue, Jul 18, 2023, 5:16 PM Aaron ***@***.***> wrote:

> In the interest of moving forward, I'd be okay if this PR did not include
> a solution for kerning
>
> Yes, we should definitely move on.
>
> I wonder now if it even makes sense to keep the chord symbols module. The
> thing it offered in V4 is 1) the ability to mix text and glyphs, and 2) the
> ability to pass in your own metrics outside of VexFlow main build. 1 is no
> longer required, because we are using the glyphs/test directly from the
> font. And 2 is no longer possible because there are no longer any metrics
> to use for formatting.
>
> So the only thing in ChordSymbol that isn't also in annotation is
> super/subscript, which we could move to annotation at some future time. It
> might be useful for other things.
>
> If you want to keep it, you can probably just go with what you have.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vexflow/vexflow/pull/15#issuecomment-1641182678>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB2MCNJIOHMYMKAOM4CSPTXQ4ROHANCNFSM6AAAAAA2EXA7S4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ronyeh commented on tests/font_tests.ts:
```diff
@@ -75,7 +74,7 @@ function fontParsing(assert: Assert): void {
   assert.equal(bFont?.style, FontStyle.NORMAL);
 
-  const f1 = 'Roboto Slab, serif';
+  const f1 = 'RobotoSlab, serif';
```
Should it be 'Roboto Slab', with a space?

rvilarl replied:
done

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -628,33 +351,12 @@ export class ChordSymbol extends Modifier {
+      symbolModifier: SymbolModifiers;
+    }> = {}
+  ): this {
+    return this.addText(String.fromCharCode(0xe874, 0xe874), params);
```
I'm beginning to think that we need comments for each SMuFL unicode codepoint that we stick in our code base. 

`0xe874` means nothing to me :-)

Maybe we should always add a comment above lines that include SMuFL code points?

```ts
// SMuFL U+E874 / csymMinor / Minor
return this.addText(String.fromCharCode(0xe874, 0xe874), params);
```

If you do it for this one, I can help out with other ones by submitting PRs. I think having this comment will help future readers of our code.

rvilarl replied:
Changed like other comments in the file with:
/*csymMinor*/

rvilarl commented:
Is there anything still to do in this one?

ronyeh commented:
I'll give it one more look and then will merge if there's nothing major. Probably by Tuesday your time.

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -9,51 +9,42 @@
-    superSubRatio: number;
-  };
-  glyphs: Record<string, ChordSymbolGlyphMetrics>;
+  isSuperscript(): boolean {
+    return this.symbolModifier !== undefined && this.symbolModifier === SymbolModifiers.SUPERSCRIPT;
+  }
+
+  isSubscript(): boolean {
+    return this.symbolModifier !== undefined && this.symbolModifier === SymbolModifiers.SUBSCRIPT;
+  }
```
In both of these functions, is the undefined check unnecessary?

Can't we just check if the this.symbolModifier === SUPERSCRIPT or SUBSCRIPT? Not sure why we need to check for undefined.

rvilarl replied:
you are right, done

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -111,166 +96,49 @@ export class ChordSymbol extends Modifier {
-    const musicFont = Tables.currentMusicFont();
-    return musicFont.lookupMetric('noteHead.minPadding');
+    return Tables.lookupMetric('NoteHead.minPadding');
   }
   /**
```
Do we need an extra empty line here?

rvilarl replied:
done

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -280,108 +148,67 @@ export class ChordSymbol extends Modifier {
-        }
-        block.width += ChordSymbol.spacingBetweenBlocks * fontSize * superSubScale;
-
         // If a subscript immediately  follows a superscript block, try to
```
This line has an extra space in `"immediately  follows"`. I know it might not be your comment, but we might as well fix it since we're here....

rvilarl replied:
done

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -395,227 +222,123 @@ export class ChordSymbol extends Modifier {
    */
   get superscriptOffset(): number {
-    return ChordSymbol.superscriptOffset * this.textFormatter.fontSizeInPixels;
+    return ChordSymbol.superscriptOffset * Font.convertSizeToPixelValue(this.textFont?.size);
```
I see you do this twice:
```
this.textFont?.size
```

Can textFont be undefined? Why are we doing the optional chaining `?.` operator here?

If it is possible for it to be undefined, it will pass undefined to `Font.convertSizeToPixelValue(...)`. I don't believe that function can accept an undefined parameter.

Can we remove the ? from the two places we use it? I thought textFont is required to be defined.

rvilarl replied:
done

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -395,227 +222,123 @@ export class ChordSymbol extends Modifier {
-      }
-    }
-    return adjustment;
+    return ChordSymbol.subscriptOffset * Font.convertSizeToPixelValue(this.textFont?.size);
```
See the other comment about optional chaining. Let's make sure we aren't accidentally passing `undefined` to Font.convertSizeToPixelValue().

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -395,227 +222,123 @@ export class ChordSymbol extends Modifier {
-    }
-  }
-
-  updateKerningAdjustments(): void {
```
I'm OK with removing the kerning code for now. If we ever decide we need it in V5, we can come back to this PR and see if the V4 approach can help us.

rvilarl replied:
ok

ronyeh commented on src/chordsymbol.ts:
```diff
@@ -395,227 +222,123 @@ export class ChordSymbol extends Modifier {
+      false
+    );
+
+    if (symbolBlock.isSubscript()) symbolBlock.setYShift(this.subscriptOffset);
+    if (symbolBlock.isSuperscript()) symbolBlock.setYShift(this.superscriptOffset);
```
In general I am not a fan of single line if statements. I'd prefer to always use braces even if it's only one statement:

```js
if (condition) {
    <a single line of code>
}
```

It sets us up for the future and prevents possible bugs (in the future).

rvilarl replied:
done

ronyeh commented on tests/chordsymbol_tests.ts:
```diff
@@ -39,18 +38,11 @@ const note = (factory: Factory, keys: string[], duration: string, chordSymbol: C
-  return widthInEm * 38 * Font.scaleToPxFrom.pt;
+function getGlyphWidth(charCode: number): number {
+  const el = new Element();
+  el.setText(String.fromCharCode(charCode));
+  el.measureText()
```
We can discuss this after this PR is merged. Should setText include `this.measureText()` as the last line of its implementation? That way, every time you setText(), the width will be valid. Is there a big performance impact if we do this?

I'm just worried since every time we call setText(...) and change the text, the width is now incorrect. A future developer (or ourselves someday, when we forget this detail), might not know that the width is now incorrect, and can only be fixed by calling measureText() again.

rvilarl replied:
There are elements adding padding. I agree that we should look into that. Perhaps an issue to resolve after the migration is complete?

ronyeh reviewed (changes requested):
I'm done reviewing, and most of my comments suggest minor changes.

I'll merge it once those are addressed. Thanks!

ronyeh commented:
Thanks!
